From 3286d7ed7dc37a587a4a6f929e03e5c3e64e2604 Mon Sep 17 00:00:00 2001
From: Rick Chen <rick@andestech.com>
Date: Wed, 18 May 2022 17:06:52 +0800
Subject: [PATCH 04/34] Support Fast Boot for Andes boards

Fast Boot is introduced to speed up the booting process, allowing
to boot a Linux kernel directly without a full blown U-Boot.i

This patch set includes the following patches:

	4a9c144530 riscv: Adjust kernel_addr_r environment variable
	7f4049a5f3 riscv: Increase SPL_SYS_MALLOC_F_LEN
	3d910df4df riscv: Disable SYS_SPL_MALLOC_START
	8471035c03 riscv: ae350: rv32 change kernel load address
	69c8eb0b69 riscv: Support QILAI board
	5e88e7503a configs: Refactor Andes default configs
	ffbc387fcc Revert "Revert "Fix data abort caused by mis-aligning FIT data""
	9c42d752b6 riscv: ae350: Support Fast Boot

Signed-off-by: Rick Chen <rick@andestech.com>
---
 Makefile                                     |  5 +-
 arch/riscv/Kconfig                           | 36 +++++++++
 arch/riscv/dts/ae350_32.dts                  |  2 +-
 arch/riscv/dts/ae350_64.dts                  |  2 +-
 arch/riscv/dts/binman.dtsi                   |  6 +-
 arch/riscv/dts/binman_d.dtsi                 | 78 ++++++++++++++++++++
 arch/riscv/dts/binman_j.dtsi                 | 78 ++++++++++++++++++++
 arch/riscv/dts/binman_p.dtsi                 | 74 +++++++++++++++++++
 board/AndesTech/ae350/Kconfig                |  4 +
 board/AndesTech/qilai/Kconfig                |  6 +-
 common/spl/spl_fit.c                         | 10 ++-
 common/spl/spl_opensbi.c                     | 20 +++--
 configs/ae350_a25_spl_defconfig              |  3 +-
 configs/ae350_a25_spl_xip_defconfig          |  3 +-
 configs/ae350_a45_spl_defconfig              |  6 +-
 configs/ae350_a45_spl_xip_defconfig          |  6 +-
 configs/ae350_ax25_spl_defconfig             |  3 +-
 configs/ae350_ax25_spl_xip_defconfig         |  3 +-
 configs/ae350_ax45_spl_defconfig             |  4 +-
 configs/ae350_ax45_spl_xip_defconfig         |  6 +-
 configs/ae350_rv32_spl_dynamic_defconfig     | 46 ++++++++++++
 configs/ae350_rv32_spl_jump_defconfig        | 46 ++++++++++++
 configs/ae350_rv32_spl_payload_defconfig     | 46 ++++++++++++
 configs/ae350_rv32_spl_xip_dynamic_defconfig | 48 ++++++++++++
 configs/ae350_rv32_spl_xip_jump_defconfig    | 48 ++++++++++++
 configs/ae350_rv32_spl_xip_payload_defconfig | 48 ++++++++++++
 configs/ae350_rv64_spl_dynamic_defconfig     | 47 ++++++++++++
 configs/ae350_rv64_spl_jump_defconfig        | 47 ++++++++++++
 configs/ae350_rv64_spl_payload_defconfig     | 47 ++++++++++++
 configs/ae350_rv64_spl_xip_dynamic_defconfig | 49 ++++++++++++
 configs/ae350_rv64_spl_xip_jump_defconfig    | 49 ++++++++++++
 configs/ae350_rv64_spl_xip_payload_defconfig | 49 ++++++++++++
 configs/qilai_a45_spl_defconfig              | 15 ++--
 configs/qilai_ax45_spl_defconfig             | 14 ++--
 include/configs/andes25-ae350.h              | 12 ++-
 include/configs/andes45-ae350.h              | 12 ++-
 include/configs/andes45-qilai.h              | 26 ++++---
 37 files changed, 928 insertions(+), 76 deletions(-)
 create mode 100644 arch/riscv/dts/binman_d.dtsi
 create mode 100644 arch/riscv/dts/binman_j.dtsi
 create mode 100644 arch/riscv/dts/binman_p.dtsi
 create mode 100644 configs/ae350_rv32_spl_dynamic_defconfig
 create mode 100644 configs/ae350_rv32_spl_jump_defconfig
 create mode 100644 configs/ae350_rv32_spl_payload_defconfig
 create mode 100644 configs/ae350_rv32_spl_xip_dynamic_defconfig
 create mode 100644 configs/ae350_rv32_spl_xip_jump_defconfig
 create mode 100644 configs/ae350_rv32_spl_xip_payload_defconfig
 create mode 100644 configs/ae350_rv64_spl_dynamic_defconfig
 create mode 100644 configs/ae350_rv64_spl_jump_defconfig
 create mode 100644 configs/ae350_rv64_spl_payload_defconfig
 create mode 100644 configs/ae350_rv64_spl_xip_dynamic_defconfig
 create mode 100644 configs/ae350_rv64_spl_xip_jump_defconfig
 create mode 100644 configs/ae350_rv64_spl_xip_payload_defconfig

diff --git a/Makefile b/Makefile
index ad83d60dc3..3702350fd6 100644
--- a/Makefile
+++ b/Makefile
@@ -2189,9 +2189,8 @@ CLEAN_DIRS  += $(MODVERDIR) \
 	       $(foreach d, spl tpl, $(patsubst %,$d/%, \
 			$(filter-out include, $(shell ls -1 $d 2>/dev/null))))
 
-CLEAN_FILES += include/bmp_logo.h include/bmp_logo_data.h \
-	       include/generated/env.in drivers/video/u_boot_logo.S \
-	       tools/version.h u-boot* MLO* SPL System.map fit-dtb.blob* \
+CLEAN_FILES += include/bmp_logo.h include/bmp_logo_data.h tools/version.h \
+	       u-boot* *.itb MLO* SPL System.map fit-dtb.blob* \
 	       u-boot-ivt.img.log u-boot-dtb.imx.log SPL.log u-boot.imx.log \
 	       lpc32xx-* bl31.c bl31.elf bl31_*.bin image.map tispl.bin* \
 	       idbloader.img flash.bin flash.log defconfig keep-syms-lto.c \
diff --git a/arch/riscv/Kconfig b/arch/riscv/Kconfig
index 4eed3b43ce..5281f93c2c 100644
--- a/arch/riscv/Kconfig
+++ b/arch/riscv/Kconfig
@@ -242,6 +242,42 @@ config NR_CPUS
 	  Stack memory is pre-allocated. U-Boot must therefore know the
 	  maximum number of CPUs that may be present.
 
+choice
+	prompt "Fit type"
+	default SPL_LOAD_FIT_NORMAL
+
+config SPL_LOAD_FIT_NORMAL
+	bool "Enable SPL applying u-boot proper from FIT"
+	depends on SPL_LOAD_FIT
+	help
+	  Use fw_dynamic from the FIT image.
+
+config SPL_LOAD_FIT_DYNAMIC
+	bool "Enable SPL (dynamic mode) applying linux from FIT"
+	depends on SPL_LOAD_FIT
+	help
+	  Use fw_dynamic from the FIT image.
+
+config SPL_LOAD_FIT_JUMP
+	bool "Enable SPL (jump mode) applying linux from FIT"
+	depends on SPL_LOAD_FIT
+	help
+	  Use fw_jump from the FIT image.
+
+config SPL_LOAD_FIT_PAYLOAD
+	bool "Enable SPL (payload mode) applying linux from FIT"
+	depends on SPL_LOAD_FIT
+	help
+	  Use fw_payload from the FIT image.
+
+endchoice
+
+config SPL_LOAD_FIT_CONFIG
+	string "Default FIT configuration"
+	depends on SPL_LOAD_FIT
+	help
+	  Specify corresponding FIT configuration for each modes.
+
 config SBI
 	bool
 	default y if RISCV_SMODE || SPL_RISCV_SMODE
diff --git a/arch/riscv/dts/ae350_32.dts b/arch/riscv/dts/ae350_32.dts
index 083f676333..3eca476c9a 100644
--- a/arch/riscv/dts/ae350_32.dts
+++ b/arch/riscv/dts/ae350_32.dts
@@ -2,7 +2,7 @@
 
 /dts-v1/;
 
-#include "binman.dtsi"
+#include CONFIG_SPL_LOAD_FIT_CONFIG
 #include "ae350-u-boot.dtsi"
 
 / {
diff --git a/arch/riscv/dts/ae350_64.dts b/arch/riscv/dts/ae350_64.dts
index 74cff9122d..3c12e01ea7 100644
--- a/arch/riscv/dts/ae350_64.dts
+++ b/arch/riscv/dts/ae350_64.dts
@@ -2,7 +2,7 @@
 
 /dts-v1/;
 
-#include "binman.dtsi"
+#include CONFIG_SPL_LOAD_FIT_CONFIG
 #include "ae350-u-boot.dtsi"
 
 / {
diff --git a/arch/riscv/dts/binman.dtsi b/arch/riscv/dts/binman.dtsi
index 5757ef65ea..b8d4ca63d2 100644
--- a/arch/riscv/dts/binman.dtsi
+++ b/arch/riscv/dts/binman.dtsi
@@ -27,7 +27,7 @@
 					os = "U-Boot";
 					arch = "riscv";
 					compression = "none";
-					load = <CONFIG_SYS_TEXT_BASE>;
+					load = /bits/64 <CONFIG_SYS_TEXT_BASE>;
 
 					uboot_blob: blob-ext {
 						filename = "u-boot-nodtb.bin";
@@ -40,8 +40,8 @@
 					os = "opensbi";
 					arch = "riscv";
 					compression = "none";
-					load = <CONFIG_SPL_OPENSBI_LOAD_ADDR>;
-					entry = <CONFIG_SPL_OPENSBI_LOAD_ADDR>;
+					load = /bits/64 <CONFIG_SPL_OPENSBI_LOAD_ADDR>;
+					entry = /bits/64 <CONFIG_SPL_OPENSBI_LOAD_ADDR>;
 
 					opensbi_blob: opensbi {
 						filename = "fw_dynamic.bin";
diff --git a/arch/riscv/dts/binman_d.dtsi b/arch/riscv/dts/binman_d.dtsi
new file mode 100644
index 0000000000..35e0220919
--- /dev/null
+++ b/arch/riscv/dts/binman_d.dtsi
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright (C) 2021, Bin Meng <bmeng.cn@gmail.com>
+ */
+
+#include <config.h>
+
+/ {
+	binman: binman {
+		multiple-images;
+	};
+};
+
+&binman {
+	itb {
+		filename = "linux.itb";
+
+		fit {
+			description = "Configuration to load OpenSBI before Linux";
+			#address-cells = <1>;
+			fit,fdt-list = "of-list";
+
+			images {
+				linux {
+					description = "Linux";
+					type = "standalone";
+					os = "Linux";
+					arch = "riscv";
+					compression = "none";
+					load = <CONFIG_SYS_TEXT_BASE>;
+
+					linux_blob: blob-ext {
+						filename = "Image";
+					};
+				};
+
+				opensbi {
+					description = "OpenSBI fw_dynamic Firmware";
+					type = "firmware";
+					os = "opensbi";
+					arch = "riscv";
+					compression = "none";
+					load = <CONFIG_SPL_OPENSBI_LOAD_ADDR>;
+					entry = <CONFIG_SPL_OPENSBI_LOAD_ADDR>;
+
+					opensbi_blob: opensbi {
+						filename = "fw_dynamic.bin";
+					};
+				};
+
+#ifndef CONFIG_OF_BOARD
+				@fdt-SEQ {
+					description = "NAME";
+					type = "flat_dt";
+					compression = "none";
+				};
+#endif
+			};
+
+			configurations {
+				default = "conf-1";
+
+#ifndef CONFIG_OF_BOARD
+				@conf-SEQ {
+#else
+				conf-1 {
+#endif
+					description = "NAME";
+					firmware = "opensbi";
+					loadables = "linux";
+#ifndef CONFIG_OF_BOARD
+					fdt = "fdt-SEQ";
+#endif
+				};
+			};
+		};
+	};
+};
diff --git a/arch/riscv/dts/binman_j.dtsi b/arch/riscv/dts/binman_j.dtsi
new file mode 100644
index 0000000000..6cbdf1f08a
--- /dev/null
+++ b/arch/riscv/dts/binman_j.dtsi
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright (C) 2021, Bin Meng <bmeng.cn@gmail.com>
+ */
+
+#include <config.h>
+
+/ {
+	binman: binman {
+		multiple-images;
+	};
+};
+
+&binman {
+	itb {
+		filename = "linux.itb";
+
+		fit {
+			description = "Configuration to load OpenSBI before Linux";
+			#address-cells = <1>;
+			fit,fdt-list = "of-list";
+
+			images {
+				linux {
+					description = "Linux";
+					type = "standalone";
+					os = "Linux";
+					arch = "riscv";
+					compression = "none";
+					load = <CONFIG_SYS_TEXT_BASE>;
+
+					linux_blob: blob-ext {
+						filename = "Image";
+					};
+				};
+
+				opensbi {
+					description = "OpenSBI fw_jump Firmware";
+					type = "firmware";
+					os = "opensbi";
+					arch = "riscv";
+					compression = "none";
+					load = <CONFIG_SPL_OPENSBI_LOAD_ADDR>;
+					entry = <CONFIG_SPL_OPENSBI_LOAD_ADDR>;
+
+					opensbi_blob: opensbi {
+						filename = "fw_jump.bin";
+					};
+				};
+
+#ifndef CONFIG_OF_BOARD
+				@fdt-SEQ {
+					description = "NAME";
+					type = "flat_dt";
+					compression = "none";
+				};
+#endif
+			};
+
+			configurations {
+				default = "conf-1";
+
+#ifndef CONFIG_OF_BOARD
+				@conf-SEQ {
+#else
+				conf-1 {
+#endif
+					description = "NAME";
+					firmware = "opensbi";
+					loadables = "linux";
+#ifndef CONFIG_OF_BOARD
+					fdt = "fdt-SEQ";
+#endif
+				};
+			};
+		};
+	};
+};
diff --git a/arch/riscv/dts/binman_p.dtsi b/arch/riscv/dts/binman_p.dtsi
new file mode 100644
index 0000000000..7bafaf8185
--- /dev/null
+++ b/arch/riscv/dts/binman_p.dtsi
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright (C) 2021, Bin Meng <bmeng.cn@gmail.com>
+ */
+
+#include <config.h>
+
+/ {
+	binman: binman {
+		multiple-images;
+	};
+};
+
+&binman {
+	itb {
+		filename = "linux.itb";
+
+		fit {
+			description = "Configuration to load OpenSBI before Linux";
+			#address-cells = <1>;
+			fit,fdt-list = "of-list";
+
+			images {
+				linux {
+					description = "Linux";
+					type = "standalone";
+					os = "Linux";
+					arch = "riscv";
+					compression = "none";
+					load = <CONFIG_SYS_TEXT_BASE>;					
+				};
+
+				opensbi {
+					description = "OpenSBI fw_payload Firmware";
+					type = "firmware";
+					os = "opensbi";
+					arch = "riscv";
+					compression = "none";
+					load = <CONFIG_SPL_OPENSBI_LOAD_ADDR>;
+					entry = <CONFIG_SPL_OPENSBI_LOAD_ADDR>;
+
+					opensbi_blob: opensbi {
+						filename = "fw_payload.bin";
+					};
+				};
+
+#ifndef CONFIG_OF_BOARD
+				@fdt-SEQ {
+					description = "NAME";
+					type = "flat_dt";
+					compression = "none";
+				};
+#endif
+			};
+
+			configurations {
+				default = "conf-1";
+
+#ifndef CONFIG_OF_BOARD
+				@conf-SEQ {
+#else
+				conf-1 {
+#endif
+					description = "NAME";
+					firmware = "opensbi";
+					loadables = "linux";
+#ifndef CONFIG_OF_BOARD
+					fdt = "fdt-SEQ";
+#endif
+				};
+			};
+		};
+	};
+};
diff --git a/board/AndesTech/ae350/Kconfig b/board/AndesTech/ae350/Kconfig
index 816da061cf..9ae69d0597 100644
--- a/board/AndesTech/ae350/Kconfig
+++ b/board/AndesTech/ae350/Kconfig
@@ -19,9 +19,11 @@ config ENV_OFFSET
 	default 0x140000 if ENV_IS_IN_SPI_FLASH
 
 config SPL_TEXT_BASE
+	default 0x02000000 if (SPL_LOAD_FIT_PAYLOAD || SPL_LOAD_FIT_JUMP)  && RISCV
 	default 0x800000
 
 config SPL_OPENSBI_LOAD_ADDR
+	default 0x00000000 if (SPL_LOAD_FIT_PAYLOAD || SPL_LOAD_FIT_JUMP)  && RISCV
 	default 0x01000000
 
 config BOARD_SPECIFIC_OPTIONS # dummy
@@ -58,9 +60,11 @@ config ENV_OFFSET
 	default 0x140000 if ENV_IS_IN_SPI_FLASH
 
 config SPL_TEXT_BASE
+	default 0x02000000 if (SPL_LOAD_FIT_PAYLOAD || SPL_LOAD_FIT_JUMP)  && RISCV
 	default 0x800000
 
 config SPL_OPENSBI_LOAD_ADDR
+	default 0x00000000 if (SPL_LOAD_FIT_PAYLOAD || SPL_LOAD_FIT_JUMP)  && RISCV
 	default 0x01000000
 
 config BOARD_SPECIFIC_OPTIONS # dummy
diff --git a/board/AndesTech/qilai/Kconfig b/board/AndesTech/qilai/Kconfig
index 7e7dba75a8..c4ad66ea89 100644
--- a/board/AndesTech/qilai/Kconfig
+++ b/board/AndesTech/qilai/Kconfig
@@ -19,10 +19,12 @@ config ENV_OFFSET
 	default 0x140000 if ENV_IS_IN_SPI_FLASH
 
 config SPL_TEXT_BASE
-	default 0x000000
+	default 0x402000000 if (SPL_LOAD_FIT_PAYLOAD || SPL_LOAD_FIT_JUMP)  && RISCV
+	default 0x400800000
 
 config SPL_OPENSBI_LOAD_ADDR
-	default 0x01000000
+	default 0x400000000 if (SPL_LOAD_FIT_PAYLOAD || SPL_LOAD_FIT_JUMP)  && RISCV
+	default 0x401000000
 
 config BOARD_SPECIFIC_OPTIONS # dummy
 	def_bool y
diff --git a/common/spl/spl_fit.c b/common/spl/spl_fit.c
index 1bbf824684..a3076e5441 100644
--- a/common/spl/spl_fit.c
+++ b/common/spl/spl_fit.c
@@ -363,7 +363,8 @@ static bool os_takes_devicetree(uint8_t os)
 	case IH_OS_U_BOOT:
 		return true;
 	case IH_OS_LINUX:
-		return IS_ENABLED(CONFIG_SPL_OS_BOOT);
+		return IS_ENABLED(CONFIG_SPL_OS_BOOT) ||
+			IS_ENABLED(CONFIG_SPL_OPENSBI);
 	default:
 		return false;
 	}
@@ -378,9 +379,12 @@ static int spl_fit_append_fdt(struct spl_image_info *spl_image,
 
 	/*
 	 * Use the address following the image as target address for the
-	 * device tree.
+	 * device tree. Load address is aligned to 8 bytes to match the required
+	 * alignment specified for linux arm [1] and arm 64 [2] booting
+	 * [1]: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/arm/booting.rst#n126
+	 * [2]: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/arm64/booting.rst#n45
 	 */
-	image_info.load_addr = spl_image->load_addr + spl_image->size;
+	image_info.load_addr = ALIGN(spl_image->load_addr + spl_image->size, 8);
 
 	/* Figure out which device tree the board wants to use */
 	node = spl_fit_get_image_node(ctx, FIT_FDT_PROP, index++);
diff --git a/common/spl/spl_opensbi.c b/common/spl/spl_opensbi.c
index d5ce9f70f0..d48304c895 100644
--- a/common/spl/spl_opensbi.c
+++ b/common/spl/spl_opensbi.c
@@ -20,7 +20,7 @@ DECLARE_GLOBAL_DATA_PTR;
 
 struct fw_dynamic_info opensbi_info;
 
-static int spl_opensbi_find_uboot_node(void *blob, int *uboot_node)
+static int spl_opensbi_find_os_node(void *blob, int *os_node)
 {
 	int fit_images_node, node;
 	const char *fit_os;
@@ -34,10 +34,8 @@ static int spl_opensbi_find_uboot_node(void *blob, int *uboot_node)
 		if (!fit_os)
 			continue;
 
-		if (genimg_get_os_id(fit_os) == IH_OS_U_BOOT) {
-			*uboot_node = node;
+		*os_node = node;
 			return 0;
-		}
 	}
 
 	return -ENODEV;
@@ -45,8 +43,8 @@ static int spl_opensbi_find_uboot_node(void *blob, int *uboot_node)
 
 void spl_invoke_opensbi(struct spl_image_info *spl_image)
 {
-	int ret, uboot_node;
-	ulong uboot_entry;
+	int ret, os_node;
+	ulong os_entry;
 	void (*opensbi_entry)(ulong hartid, ulong dtb, ulong info);
 
 	if (!spl_image->fdt_addr) {
@@ -54,22 +52,22 @@ void spl_invoke_opensbi(struct spl_image_info *spl_image)
 		hang();
 	}
 
-	/* Find U-Boot image in /fit-images */
-	ret = spl_opensbi_find_uboot_node(spl_image->fdt_addr, &uboot_node);
+	/* Find os image, U-Boot or Linux image in /fit-images */
+	ret = spl_opensbi_find_os_node(spl_image->fdt_addr, &os_node);
 	if (ret) {
 		pr_err("Can't find U-Boot node, %d\n", ret);
 		hang();
 	}
 
 	/* Get U-Boot entry point */
-	ret = fit_image_get_entry(spl_image->fdt_addr, uboot_node, &uboot_entry);
+	ret = fit_image_get_entry(spl_image->fdt_addr, os_node, &os_entry);
 	if (ret)
-		ret = fit_image_get_load(spl_image->fdt_addr, uboot_node, &uboot_entry);
+		ret = fit_image_get_load(spl_image->fdt_addr, os_node, &os_entry);
 
 	/* Prepare obensbi_info object */
 	opensbi_info.magic = FW_DYNAMIC_INFO_MAGIC_VALUE;
 	opensbi_info.version = FW_DYNAMIC_INFO_VERSION;
-	opensbi_info.next_addr = uboot_entry;
+	opensbi_info.next_addr = os_entry;
 	opensbi_info.next_mode = FW_DYNAMIC_INFO_NEXT_MODE_S;
 	opensbi_info.boot_hart = gd->arch.boot_hart;
 
diff --git a/configs/ae350_a25_spl_defconfig b/configs/ae350_a25_spl_defconfig
index 660fc8ebe4..951a49030f 100644
--- a/configs/ae350_a25_spl_defconfig
+++ b/configs/ae350_a25_spl_defconfig
@@ -6,9 +6,9 @@ CONFIG_ENV_SECT_SIZE=0x1000
 CONFIG_DEFAULT_DEVICE_TREE="ae350_32"
 CONFIG_SPL_SYS_MALLOC_F_LEN=0x100000
 CONFIG_SPL=y
-CONFIG_SYS_LOAD_ADDR=0x100000
 CONFIG_TARGET_ANDES25_AE350=y
 CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman.dtsi"
 CONFIG_DISTRO_DEFAULTS=y
 CONFIG_SYS_LOAD_ADDR=0x100000
 CONFIG_FIT=y
@@ -23,7 +23,6 @@ CONFIG_CMD_SF_TEST=y
 # CONFIG_CMD_SETEXPR is not set
 CONFIG_BOOTP_PREFER_SERVERIP=y
 CONFIG_CMD_CACHE=y
-CONFIG_OF_BOARD=y
 CONFIG_ENV_OVERWRITE=y
 CONFIG_ENV_IS_IN_SPI_FLASH=y
 CONFIG_BOOTP_SEND_HOSTNAME=y
diff --git a/configs/ae350_a25_spl_xip_defconfig b/configs/ae350_a25_spl_xip_defconfig
index af18316a01..a20cfbfd87 100644
--- a/configs/ae350_a25_spl_xip_defconfig
+++ b/configs/ae350_a25_spl_xip_defconfig
@@ -7,9 +7,9 @@ CONFIG_DEFAULT_DEVICE_TREE="ae350_32"
 CONFIG_SPL_TEXT_BASE=0x80000000
 CONFIG_SPL_SYS_MALLOC_F_LEN=0x100000
 CONFIG_SPL=y
-CONFIG_SYS_LOAD_ADDR=0x100000
 CONFIG_TARGET_ANDES25_AE350=y
 CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman.dtsi"
 CONFIG_XIP=y
 CONFIG_DISTRO_DEFAULTS=y
 CONFIG_SYS_LOAD_ADDR=0x100000
@@ -25,7 +25,6 @@ CONFIG_CMD_SF_TEST=y
 # CONFIG_CMD_SETEXPR is not set
 CONFIG_BOOTP_PREFER_SERVERIP=y
 CONFIG_CMD_CACHE=y
-CONFIG_OF_BOARD=y
 CONFIG_ENV_OVERWRITE=y
 CONFIG_ENV_IS_IN_SPI_FLASH=y
 CONFIG_BOOTP_SEND_HOSTNAME=y
diff --git a/configs/ae350_a45_spl_defconfig b/configs/ae350_a45_spl_defconfig
index dfbed11cdc..8bac49c4a3 100644
--- a/configs/ae350_a45_spl_defconfig
+++ b/configs/ae350_a45_spl_defconfig
@@ -6,16 +6,17 @@ CONFIG_ENV_SECT_SIZE=0x1000
 CONFIG_DEFAULT_DEVICE_TREE="ae350_32"
 CONFIG_SPL_SYS_MALLOC_F_LEN=0x100000
 CONFIG_SPL=y
-CONFIG_SYS_LOAD_ADDR=0x100000
 CONFIG_TARGET_ANDES45_AE350=y
 CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman.dtsi"
 CONFIG_DISTRO_DEFAULTS=y
 CONFIG_PHYS_64BIT=y
+CONFIG_SYS_LOAD_ADDR=0x100000
 CONFIG_FIT=y
 CONFIG_SPL_LOAD_FIT_ADDRESS=0x00200000
-CONFIG_SYS_MONITOR_BASE=0x88000000
 CONFIG_BOOTDELAY=3
 CONFIG_MISC_INIT_R=y
+CONFIG_SPL_BOARD_INIT=y
 CONFIG_SPL_CACHE=y
 CONFIG_SYS_PROMPT="RISC-V # "
 CONFIG_CMD_IMLS=y
@@ -24,7 +25,6 @@ CONFIG_CMD_SF_TEST=y
 # CONFIG_CMD_SETEXPR is not set
 CONFIG_BOOTP_PREFER_SERVERIP=y
 CONFIG_CMD_CACHE=y
-CONFIG_OF_BOARD=y
 CONFIG_OF_LIST="ae350_64"
 CONFIG_ENV_OVERWRITE=y
 CONFIG_ENV_IS_IN_SPI_FLASH=y
diff --git a/configs/ae350_a45_spl_xip_defconfig b/configs/ae350_a45_spl_xip_defconfig
index 75c82b264f..02bad98006 100644
--- a/configs/ae350_a45_spl_xip_defconfig
+++ b/configs/ae350_a45_spl_xip_defconfig
@@ -7,17 +7,18 @@ CONFIG_DEFAULT_DEVICE_TREE="ae350_32"
 CONFIG_SPL_TEXT_BASE=0x80000000
 CONFIG_SPL_SYS_MALLOC_F_LEN=0x100000
 CONFIG_SPL=y
-CONFIG_SYS_LOAD_ADDR=0x100000
 CONFIG_TARGET_ANDES45_AE350=y
 CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman.dtsi"
 CONFIG_XIP=y
 CONFIG_DISTRO_DEFAULTS=y
 CONFIG_PHYS_64BIT=y
+CONFIG_SYS_LOAD_ADDR=0x100000
 CONFIG_FIT=y
 CONFIG_SPL_LOAD_FIT_ADDRESS=0x80010000
-CONFIG_SYS_MONITOR_BASE=0x88000000
 CONFIG_BOOTDELAY=3
 CONFIG_MISC_INIT_R=y
+CONFIG_SPL_BOARD_INIT=y
 CONFIG_SPL_CACHE=y
 CONFIG_SYS_PROMPT="RISC-V # "
 CONFIG_CMD_IMLS=y
@@ -26,7 +27,6 @@ CONFIG_CMD_SF_TEST=y
 # CONFIG_CMD_SETEXPR is not set
 CONFIG_BOOTP_PREFER_SERVERIP=y
 CONFIG_CMD_CACHE=y
-CONFIG_OF_BOARD=y
 CONFIG_OF_LIST="ae350_64"
 CONFIG_ENV_OVERWRITE=y
 CONFIG_ENV_IS_IN_SPI_FLASH=y
diff --git a/configs/ae350_ax25_spl_defconfig b/configs/ae350_ax25_spl_defconfig
index 85c2878045..c2ae322dbb 100644
--- a/configs/ae350_ax25_spl_defconfig
+++ b/configs/ae350_ax25_spl_defconfig
@@ -6,13 +6,13 @@ CONFIG_ENV_SECT_SIZE=0x1000
 CONFIG_DEFAULT_DEVICE_TREE="ae350_64"
 CONFIG_SPL_SYS_MALLOC_F_LEN=0x100000
 CONFIG_SPL=y
-CONFIG_SYS_LOAD_ADDR=0x100000
 CONFIG_TARGET_ANDES25_AE350=y
 CONFIG_ARCH_RV64I=y
 CONFIG_RISCV_SMODE=y
 CONFIG_DISTRO_DEFAULTS=y
 CONFIG_SYS_LOAD_ADDR=0x100000
 CONFIG_FIT=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman.dtsi"
 CONFIG_SPL_LOAD_FIT_ADDRESS=0x00200000
 CONFIG_BOOTDELAY=3
 CONFIG_MISC_INIT_R=y
@@ -24,7 +24,6 @@ CONFIG_CMD_SF_TEST=y
 # CONFIG_CMD_SETEXPR is not set
 CONFIG_BOOTP_PREFER_SERVERIP=y
 CONFIG_CMD_CACHE=y
-CONFIG_OF_BOARD=y
 CONFIG_ENV_OVERWRITE=y
 CONFIG_ENV_IS_IN_SPI_FLASH=y
 CONFIG_BOOTP_SEND_HOSTNAME=y
diff --git a/configs/ae350_ax25_spl_xip_defconfig b/configs/ae350_ax25_spl_xip_defconfig
index 8ba3376a5f..8ac8d5d64a 100644
--- a/configs/ae350_ax25_spl_xip_defconfig
+++ b/configs/ae350_ax25_spl_xip_defconfig
@@ -7,10 +7,10 @@ CONFIG_DEFAULT_DEVICE_TREE="ae350_64"
 CONFIG_SPL_TEXT_BASE=0x80000000
 CONFIG_SPL_SYS_MALLOC_F_LEN=0x100000
 CONFIG_SPL=y
-CONFIG_SYS_LOAD_ADDR=0x100000
 CONFIG_TARGET_ANDES25_AE350=y
 CONFIG_ARCH_RV64I=y
 CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman.dtsi"
 CONFIG_XIP=y
 CONFIG_DISTRO_DEFAULTS=y
 CONFIG_SYS_LOAD_ADDR=0x100000
@@ -26,7 +26,6 @@ CONFIG_CMD_SF_TEST=y
 # CONFIG_CMD_SETEXPR is not set
 CONFIG_BOOTP_PREFER_SERVERIP=y
 CONFIG_CMD_CACHE=y
-CONFIG_OF_BOARD=y
 CONFIG_ENV_OVERWRITE=y
 CONFIG_ENV_IS_IN_SPI_FLASH=y
 CONFIG_BOOTP_SEND_HOSTNAME=y
diff --git a/configs/ae350_ax45_spl_defconfig b/configs/ae350_ax45_spl_defconfig
index 6b59708621..a6a320d8e7 100644
--- a/configs/ae350_ax45_spl_defconfig
+++ b/configs/ae350_ax45_spl_defconfig
@@ -9,13 +9,14 @@ CONFIG_SPL=y
 CONFIG_TARGET_ANDES45_AE350=y
 CONFIG_ARCH_RV64I=y
 CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman.dtsi"
 CONFIG_DISTRO_DEFAULTS=y
 CONFIG_SYS_LOAD_ADDR=0x100000
 CONFIG_FIT=y
 CONFIG_SPL_LOAD_FIT_ADDRESS=0x00200000
-CONFIG_SYS_MONITOR_BASE=0x88000000
 CONFIG_BOOTDELAY=3
 CONFIG_MISC_INIT_R=y
+CONFIG_SPL_BOARD_INIT=y
 CONFIG_SPL_CACHE=y
 CONFIG_SYS_PROMPT="RISC-V # "
 CONFIG_CMD_IMLS=y
@@ -24,7 +25,6 @@ CONFIG_CMD_SF_TEST=y
 # CONFIG_CMD_SETEXPR is not set
 CONFIG_BOOTP_PREFER_SERVERIP=y
 CONFIG_CMD_CACHE=y
-CONFIG_OF_BOARD=y
 CONFIG_ENV_OVERWRITE=y
 CONFIG_ENV_IS_IN_SPI_FLASH=y
 CONFIG_BOOTP_SEND_HOSTNAME=y
diff --git a/configs/ae350_ax45_spl_xip_defconfig b/configs/ae350_ax45_spl_xip_defconfig
index 9903ae8340..4125da8e25 100644
--- a/configs/ae350_ax45_spl_xip_defconfig
+++ b/configs/ae350_ax45_spl_xip_defconfig
@@ -7,17 +7,18 @@ CONFIG_DEFAULT_DEVICE_TREE="ae350_64"
 CONFIG_SPL_TEXT_BASE=0x80000000
 CONFIG_SPL_SYS_MALLOC_F_LEN=0x100000
 CONFIG_SPL=y
-CONFIG_SYS_LOAD_ADDR=0x100000
 CONFIG_TARGET_ANDES45_AE350=y
 CONFIG_ARCH_RV64I=y
 CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman.dtsi"
 CONFIG_XIP=y
 CONFIG_DISTRO_DEFAULTS=y
+CONFIG_SYS_LOAD_ADDR=0x100000
 CONFIG_FIT=y
 CONFIG_SPL_LOAD_FIT_ADDRESS=0x80010000
-CONFIG_SYS_MONITOR_BASE=0x88000000
 CONFIG_BOOTDELAY=3
 CONFIG_MISC_INIT_R=y
+CONFIG_SPL_BOARD_INIT=y
 CONFIG_SPL_CACHE=y
 CONFIG_SYS_PROMPT="RISC-V # "
 CONFIG_CMD_IMLS=y
@@ -26,7 +27,6 @@ CONFIG_CMD_SF_TEST=y
 # CONFIG_CMD_SETEXPR is not set
 CONFIG_BOOTP_PREFER_SERVERIP=y
 CONFIG_CMD_CACHE=y
-CONFIG_OF_BOARD=y
 CONFIG_ENV_OVERWRITE=y
 CONFIG_ENV_IS_IN_SPI_FLASH=y
 CONFIG_BOOTP_SEND_HOSTNAME=y
diff --git a/configs/ae350_rv32_spl_dynamic_defconfig b/configs/ae350_rv32_spl_dynamic_defconfig
new file mode 100644
index 0000000000..4c96a94576
--- /dev/null
+++ b/configs/ae350_rv32_spl_dynamic_defconfig
@@ -0,0 +1,46 @@
+CONFIG_RISCV=y
+CONFIG_SYS_TEXT_BASE=0x01800000
+CONFIG_SYS_MALLOC_LEN=0x80000
+CONFIG_NR_DRAM_BANKS=2
+CONFIG_ENV_SECT_SIZE=0x1000
+CONFIG_DEFAULT_DEVICE_TREE="ae350_32"
+CONFIG_SPL_SYS_MALLOC_F_LEN=0x1800000
+CONFIG_SPL=y
+CONFIG_TARGET_ANDES45_AE350=y
+CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_DYNAMIC=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman_d.dtsi"
+CONFIG_DISTRO_DEFAULTS=y
+CONFIG_SYS_LOAD_ADDR=0x100000
+CONFIG_FIT=y
+CONFIG_SPL_LOAD_FIT_ADDRESS=0x06000000
+CONFIG_BOOTDELAY=3
+CONFIG_MISC_INIT_R=y
+CONFIG_SPL_BOARD_INIT=y
+CONFIG_SPL_CACHE=y
+CONFIG_SYS_PROMPT="RISC-V # "
+CONFIG_CMD_IMLS=y
+CONFIG_CMD_MMC=y
+CONFIG_CMD_SF_TEST=y
+# CONFIG_CMD_SETEXPR is not set
+CONFIG_BOOTP_PREFER_SERVERIP=y
+CONFIG_CMD_CACHE=y
+CONFIG_ENV_OVERWRITE=y
+CONFIG_ENV_IS_IN_SPI_FLASH=y
+CONFIG_BOOTP_SEND_HOSTNAME=y
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_V5L2_CACHE=y
+CONFIG_MMC=y
+CONFIG_FTSDC010=y
+CONFIG_FTSDC010_SDIO=y
+CONFIG_MTD_NOR_FLASH=y
+CONFIG_FLASH_CFI_DRIVER=y
+CONFIG_SYS_FLASH_USE_BUFFER_WRITE=y
+CONFIG_SYS_FLASH_CFI=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_FTMAC100=y
+CONFIG_BAUDRATE=38400
+CONFIG_SYS_NS16550=y
+CONFIG_SPI=y
+CONFIG_ATCSPI200_SPI=y
+# CONFIG_BINMAN_FDT is not set
diff --git a/configs/ae350_rv32_spl_jump_defconfig b/configs/ae350_rv32_spl_jump_defconfig
new file mode 100644
index 0000000000..4212100556
--- /dev/null
+++ b/configs/ae350_rv32_spl_jump_defconfig
@@ -0,0 +1,46 @@
+CONFIG_RISCV=y
+CONFIG_SYS_TEXT_BASE=0x0200000
+CONFIG_SYS_MALLOC_LEN=0x80000
+CONFIG_NR_DRAM_BANKS=2
+CONFIG_ENV_SECT_SIZE=0x1000
+CONFIG_DEFAULT_DEVICE_TREE="ae350_32"
+CONFIG_SPL_SYS_MALLOC_F_LEN=0x1800000
+CONFIG_SPL=y
+CONFIG_TARGET_ANDES45_AE350=y
+CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_JUMP=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman_j.dtsi"
+CONFIG_DISTRO_DEFAULTS=y
+CONFIG_SYS_LOAD_ADDR=0x100000
+CONFIG_FIT=y
+CONFIG_SPL_LOAD_FIT_ADDRESS=0x06000000
+CONFIG_BOOTDELAY=3
+CONFIG_MISC_INIT_R=y
+CONFIG_SPL_BOARD_INIT=y
+CONFIG_SPL_CACHE=y
+CONFIG_SYS_PROMPT="RISC-V # "
+CONFIG_CMD_IMLS=y
+CONFIG_CMD_MMC=y
+CONFIG_CMD_SF_TEST=y
+# CONFIG_CMD_SETEXPR is not set
+CONFIG_BOOTP_PREFER_SERVERIP=y
+CONFIG_CMD_CACHE=y
+CONFIG_ENV_OVERWRITE=y
+CONFIG_ENV_IS_IN_SPI_FLASH=y
+CONFIG_BOOTP_SEND_HOSTNAME=y
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_V5L2_CACHE=y
+CONFIG_MMC=y
+CONFIG_FTSDC010=y
+CONFIG_FTSDC010_SDIO=y
+CONFIG_MTD_NOR_FLASH=y
+CONFIG_FLASH_CFI_DRIVER=y
+CONFIG_SYS_FLASH_USE_BUFFER_WRITE=y
+CONFIG_SYS_FLASH_CFI=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_FTMAC100=y
+CONFIG_BAUDRATE=38400
+CONFIG_SYS_NS16550=y
+CONFIG_SPI=y
+CONFIG_ATCSPI200_SPI=y
+# CONFIG_BINMAN_FDT is not set
diff --git a/configs/ae350_rv32_spl_payload_defconfig b/configs/ae350_rv32_spl_payload_defconfig
new file mode 100644
index 0000000000..fab72cde4c
--- /dev/null
+++ b/configs/ae350_rv32_spl_payload_defconfig
@@ -0,0 +1,46 @@
+CONFIG_RISCV=y
+CONFIG_SYS_TEXT_BASE=0x01200000
+CONFIG_SYS_MALLOC_LEN=0x80000
+CONFIG_NR_DRAM_BANKS=2
+CONFIG_ENV_SECT_SIZE=0x1000
+CONFIG_DEFAULT_DEVICE_TREE="ae350_32"
+CONFIG_SPL_SYS_MALLOC_F_LEN=0x1800000
+CONFIG_SPL=y
+CONFIG_TARGET_ANDES45_AE350=y
+CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_PAYLOAD=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman_p.dtsi"
+CONFIG_DISTRO_DEFAULTS=y
+CONFIG_SYS_LOAD_ADDR=0x100000
+CONFIG_FIT=y
+CONFIG_SPL_LOAD_FIT_ADDRESS=0x00200000
+CONFIG_BOOTDELAY=3
+CONFIG_MISC_INIT_R=y
+CONFIG_SPL_BOARD_INIT=y
+CONFIG_SPL_CACHE=y
+CONFIG_SYS_PROMPT="RISC-V # "
+CONFIG_CMD_IMLS=y
+CONFIG_CMD_MMC=y
+CONFIG_CMD_SF_TEST=y
+# CONFIG_CMD_SETEXPR is not set
+CONFIG_BOOTP_PREFER_SERVERIP=y
+CONFIG_CMD_CACHE=y
+CONFIG_ENV_OVERWRITE=y
+CONFIG_ENV_IS_IN_SPI_FLASH=y
+CONFIG_BOOTP_SEND_HOSTNAME=y
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_V5L2_CACHE=y
+CONFIG_MMC=y
+CONFIG_FTSDC010=y
+CONFIG_FTSDC010_SDIO=y
+CONFIG_MTD_NOR_FLASH=y
+CONFIG_FLASH_CFI_DRIVER=y
+CONFIG_SYS_FLASH_USE_BUFFER_WRITE=y
+CONFIG_SYS_FLASH_CFI=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_FTMAC100=y
+CONFIG_BAUDRATE=38400
+CONFIG_SYS_NS16550=y
+CONFIG_SPI=y
+CONFIG_ATCSPI200_SPI=y
+# CONFIG_BINMAN_FDT is not set
diff --git a/configs/ae350_rv32_spl_xip_dynamic_defconfig b/configs/ae350_rv32_spl_xip_dynamic_defconfig
new file mode 100644
index 0000000000..f7a7502c92
--- /dev/null
+++ b/configs/ae350_rv32_spl_xip_dynamic_defconfig
@@ -0,0 +1,48 @@
+CONFIG_RISCV=y
+CONFIG_SYS_TEXT_BASE=0x01200000
+CONFIG_SYS_MALLOC_LEN=0x80000
+CONFIG_NR_DRAM_BANKS=2
+CONFIG_ENV_SECT_SIZE=0x1000
+CONFIG_DEFAULT_DEVICE_TREE="ae350_32"
+CONFIG_SPL_TEXT_BASE=0x80000000
+CONFIG_SPL_SYS_MALLOC_F_LEN=0x1800000
+CONFIG_SPL=y
+CONFIG_TARGET_ANDES45_AE350=y
+CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_DYNAMIC=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman_d.dtsi"
+CONFIG_XIP=y
+CONFIG_DISTRO_DEFAULTS=y
+CONFIG_SYS_LOAD_ADDR=0x100000
+CONFIG_FIT=y
+CONFIG_SPL_LOAD_FIT_ADDRESS=0x88600000
+CONFIG_BOOTDELAY=3
+CONFIG_MISC_INIT_R=y
+CONFIG_SPL_BOARD_INIT=y
+CONFIG_SPL_CACHE=y
+CONFIG_SYS_PROMPT="RISC-V # "
+CONFIG_CMD_IMLS=y
+CONFIG_CMD_MMC=y
+CONFIG_CMD_SF_TEST=y
+# CONFIG_CMD_SETEXPR is not set
+CONFIG_BOOTP_PREFER_SERVERIP=y
+CONFIG_CMD_CACHE=y
+CONFIG_ENV_OVERWRITE=y
+CONFIG_ENV_IS_IN_SPI_FLASH=y
+CONFIG_BOOTP_SEND_HOSTNAME=y
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_V5L2_CACHE=y
+CONFIG_MMC=y
+CONFIG_FTSDC010=y
+CONFIG_FTSDC010_SDIO=y
+CONFIG_MTD_NOR_FLASH=y
+CONFIG_FLASH_CFI_DRIVER=y
+CONFIG_SYS_FLASH_USE_BUFFER_WRITE=y
+CONFIG_SYS_FLASH_CFI=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_FTMAC100=y
+CONFIG_BAUDRATE=38400
+CONFIG_SYS_NS16550=y
+CONFIG_SPI=y
+CONFIG_ATCSPI200_SPI=y
+# CONFIG_BINMAN_FDT is not set
diff --git a/configs/ae350_rv32_spl_xip_jump_defconfig b/configs/ae350_rv32_spl_xip_jump_defconfig
new file mode 100644
index 0000000000..82104ccffe
--- /dev/null
+++ b/configs/ae350_rv32_spl_xip_jump_defconfig
@@ -0,0 +1,48 @@
+CONFIG_RISCV=y
+CONFIG_SYS_TEXT_BASE=0x0200000
+CONFIG_SYS_MALLOC_LEN=0x80000
+CONFIG_NR_DRAM_BANKS=2
+CONFIG_ENV_SECT_SIZE=0x1000
+CONFIG_DEFAULT_DEVICE_TREE="ae350_32"
+CONFIG_SPL_TEXT_BASE=0x80000000
+CONFIG_SPL_SYS_MALLOC_F_LEN=0x1800000
+CONFIG_SPL=y
+CONFIG_TARGET_ANDES45_AE350=y
+CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_JUMP=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman_j.dtsi"
+CONFIG_XIP=y
+CONFIG_DISTRO_DEFAULTS=y
+CONFIG_SYS_LOAD_ADDR=0x100000
+CONFIG_FIT=y
+CONFIG_SPL_LOAD_FIT_ADDRESS=0x88600000
+CONFIG_BOOTDELAY=3
+CONFIG_MISC_INIT_R=y
+CONFIG_SPL_BOARD_INIT=y
+CONFIG_SPL_CACHE=y
+CONFIG_SYS_PROMPT="RISC-V # "
+CONFIG_CMD_IMLS=y
+CONFIG_CMD_MMC=y
+CONFIG_CMD_SF_TEST=y
+# CONFIG_CMD_SETEXPR is not set
+CONFIG_BOOTP_PREFER_SERVERIP=y
+CONFIG_CMD_CACHE=y
+CONFIG_ENV_OVERWRITE=y
+CONFIG_ENV_IS_IN_SPI_FLASH=y
+CONFIG_BOOTP_SEND_HOSTNAME=y
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_V5L2_CACHE=y
+CONFIG_MMC=y
+CONFIG_FTSDC010=y
+CONFIG_FTSDC010_SDIO=y
+CONFIG_MTD_NOR_FLASH=y
+CONFIG_FLASH_CFI_DRIVER=y
+CONFIG_SYS_FLASH_USE_BUFFER_WRITE=y
+CONFIG_SYS_FLASH_CFI=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_FTMAC100=y
+CONFIG_BAUDRATE=38400
+CONFIG_SYS_NS16550=y
+CONFIG_SPI=y
+CONFIG_ATCSPI200_SPI=y
+# CONFIG_BINMAN_FDT is not set
diff --git a/configs/ae350_rv32_spl_xip_payload_defconfig b/configs/ae350_rv32_spl_xip_payload_defconfig
new file mode 100644
index 0000000000..23da552f13
--- /dev/null
+++ b/configs/ae350_rv32_spl_xip_payload_defconfig
@@ -0,0 +1,48 @@
+CONFIG_RISCV=y
+CONFIG_SYS_TEXT_BASE=0x01200000
+CONFIG_SYS_MALLOC_LEN=0x80000
+CONFIG_NR_DRAM_BANKS=2
+CONFIG_ENV_SECT_SIZE=0x1000
+CONFIG_DEFAULT_DEVICE_TREE="ae350_32"
+CONFIG_SPL_TEXT_BASE=0x80000000
+CONFIG_SPL_SYS_MALLOC_F_LEN=0x1800000
+CONFIG_SPL=y
+CONFIG_TARGET_ANDES45_AE350=y
+CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_PAYLOAD=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman_p.dtsi"
+CONFIG_XIP=y
+CONFIG_DISTRO_DEFAULTS=y
+CONFIG_SYS_LOAD_ADDR=0x100000
+CONFIG_FIT=y
+CONFIG_SPL_LOAD_FIT_ADDRESS=0x88600000
+CONFIG_BOOTDELAY=3
+CONFIG_MISC_INIT_R=y
+CONFIG_SPL_BOARD_INIT=y
+CONFIG_SPL_CACHE=y
+CONFIG_SYS_PROMPT="RISC-V # "
+CONFIG_CMD_IMLS=y
+CONFIG_CMD_MMC=y
+CONFIG_CMD_SF_TEST=y
+# CONFIG_CMD_SETEXPR is not set
+CONFIG_BOOTP_PREFER_SERVERIP=y
+CONFIG_CMD_CACHE=y
+CONFIG_ENV_OVERWRITE=y
+CONFIG_ENV_IS_IN_SPI_FLASH=y
+CONFIG_BOOTP_SEND_HOSTNAME=y
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_V5L2_CACHE=y
+CONFIG_MMC=y
+CONFIG_FTSDC010=y
+CONFIG_FTSDC010_SDIO=y
+CONFIG_MTD_NOR_FLASH=y
+CONFIG_FLASH_CFI_DRIVER=y
+CONFIG_SYS_FLASH_USE_BUFFER_WRITE=y
+CONFIG_SYS_FLASH_CFI=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_FTMAC100=y
+CONFIG_BAUDRATE=38400
+CONFIG_SYS_NS16550=y
+CONFIG_SPI=y
+CONFIG_ATCSPI200_SPI=y
+# CONFIG_BINMAN_FDT is not set
diff --git a/configs/ae350_rv64_spl_dynamic_defconfig b/configs/ae350_rv64_spl_dynamic_defconfig
new file mode 100644
index 0000000000..3f0a9e16b7
--- /dev/null
+++ b/configs/ae350_rv64_spl_dynamic_defconfig
@@ -0,0 +1,47 @@
+CONFIG_RISCV=y
+CONFIG_SYS_TEXT_BASE=0x01200000
+CONFIG_SYS_MALLOC_LEN=0x80000
+CONFIG_NR_DRAM_BANKS=2
+CONFIG_ENV_SECT_SIZE=0x1000
+CONFIG_DEFAULT_DEVICE_TREE="ae350_64"
+CONFIG_SPL_SYS_MALLOC_F_LEN=0x1800000
+CONFIG_SPL=y
+CONFIG_TARGET_ANDES45_AE350=y
+CONFIG_ARCH_RV64I=y
+CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_DYNAMIC=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman_d.dtsi"
+CONFIG_DISTRO_DEFAULTS=y
+CONFIG_SYS_LOAD_ADDR=0x100000
+CONFIG_FIT=y
+CONFIG_SPL_LOAD_FIT_ADDRESS=0x06000000
+CONFIG_BOOTDELAY=3
+CONFIG_MISC_INIT_R=y
+CONFIG_SPL_BOARD_INIT=y
+CONFIG_SPL_CACHE=y
+CONFIG_SYS_PROMPT="RISC-V # "
+CONFIG_CMD_IMLS=y
+CONFIG_CMD_MMC=y
+CONFIG_CMD_SF_TEST=y
+# CONFIG_CMD_SETEXPR is not set
+CONFIG_BOOTP_PREFER_SERVERIP=y
+CONFIG_CMD_CACHE=y
+CONFIG_ENV_OVERWRITE=y
+CONFIG_ENV_IS_IN_SPI_FLASH=y
+CONFIG_BOOTP_SEND_HOSTNAME=y
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_V5L2_CACHE=y
+CONFIG_MMC=y
+CONFIG_FTSDC010=y
+CONFIG_FTSDC010_SDIO=y
+CONFIG_MTD_NOR_FLASH=y
+CONFIG_FLASH_CFI_DRIVER=y
+CONFIG_SYS_FLASH_USE_BUFFER_WRITE=y
+CONFIG_SYS_FLASH_CFI=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_FTMAC100=y
+CONFIG_BAUDRATE=38400
+CONFIG_SYS_NS16550=y
+CONFIG_SPI=y
+CONFIG_ATCSPI200_SPI=y
+# CONFIG_BINMAN_FDT is not set
diff --git a/configs/ae350_rv64_spl_jump_defconfig b/configs/ae350_rv64_spl_jump_defconfig
new file mode 100644
index 0000000000..c9d5a2763f
--- /dev/null
+++ b/configs/ae350_rv64_spl_jump_defconfig
@@ -0,0 +1,47 @@
+CONFIG_RISCV=y
+CONFIG_SYS_TEXT_BASE=0x0200000
+CONFIG_SYS_MALLOC_LEN=0x80000
+CONFIG_NR_DRAM_BANKS=2
+CONFIG_ENV_SECT_SIZE=0x1000
+CONFIG_DEFAULT_DEVICE_TREE="ae350_64"
+CONFIG_SPL_SYS_MALLOC_F_LEN=0x1800000
+CONFIG_SPL=y
+CONFIG_TARGET_ANDES45_AE350=y
+CONFIG_ARCH_RV64I=y
+CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_JUMP=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman_j.dtsi"
+CONFIG_DISTRO_DEFAULTS=y
+CONFIG_SYS_LOAD_ADDR=0x100000
+CONFIG_FIT=y
+CONFIG_SPL_LOAD_FIT_ADDRESS=0x06000000
+CONFIG_BOOTDELAY=3
+CONFIG_MISC_INIT_R=y
+CONFIG_SPL_BOARD_INIT=y
+CONFIG_SPL_CACHE=y
+CONFIG_SYS_PROMPT="RISC-V # "
+CONFIG_CMD_IMLS=y
+CONFIG_CMD_MMC=y
+CONFIG_CMD_SF_TEST=y
+# CONFIG_CMD_SETEXPR is not set
+CONFIG_BOOTP_PREFER_SERVERIP=y
+CONFIG_CMD_CACHE=y
+CONFIG_ENV_OVERWRITE=y
+CONFIG_ENV_IS_IN_SPI_FLASH=y
+CONFIG_BOOTP_SEND_HOSTNAME=y
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_V5L2_CACHE=y
+CONFIG_MMC=y
+CONFIG_FTSDC010=y
+CONFIG_FTSDC010_SDIO=y
+CONFIG_MTD_NOR_FLASH=y
+CONFIG_FLASH_CFI_DRIVER=y
+CONFIG_SYS_FLASH_USE_BUFFER_WRITE=y
+CONFIG_SYS_FLASH_CFI=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_FTMAC100=y
+CONFIG_BAUDRATE=38400
+CONFIG_SYS_NS16550=y
+CONFIG_SPI=y
+CONFIG_ATCSPI200_SPI=y
+# CONFIG_BINMAN_FDT is not set
diff --git a/configs/ae350_rv64_spl_payload_defconfig b/configs/ae350_rv64_spl_payload_defconfig
new file mode 100644
index 0000000000..4d24e11b29
--- /dev/null
+++ b/configs/ae350_rv64_spl_payload_defconfig
@@ -0,0 +1,47 @@
+CONFIG_RISCV=y
+CONFIG_SYS_TEXT_BASE=0x01200000
+CONFIG_SYS_MALLOC_LEN=0x80000
+CONFIG_NR_DRAM_BANKS=2
+CONFIG_ENV_SECT_SIZE=0x1000
+CONFIG_DEFAULT_DEVICE_TREE="ae350_64"
+CONFIG_SPL_SYS_MALLOC_F_LEN=0x1800000
+CONFIG_SPL=y
+CONFIG_TARGET_ANDES45_AE350=y
+CONFIG_ARCH_RV64I=y
+CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_PAYLOAD=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman_p.dtsi"
+CONFIG_DISTRO_DEFAULTS=y
+CONFIG_SYS_LOAD_ADDR=0x100000
+CONFIG_FIT=y
+CONFIG_SPL_LOAD_FIT_ADDRESS=0x00200000
+CONFIG_BOOTDELAY=3
+CONFIG_MISC_INIT_R=y
+CONFIG_SPL_BOARD_INIT=y
+CONFIG_SPL_CACHE=y
+CONFIG_SYS_PROMPT="RISC-V # "
+CONFIG_CMD_IMLS=y
+CONFIG_CMD_MMC=y
+CONFIG_CMD_SF_TEST=y
+# CONFIG_CMD_SETEXPR is not set
+CONFIG_BOOTP_PREFER_SERVERIP=y
+CONFIG_CMD_CACHE=y
+CONFIG_ENV_OVERWRITE=y
+CONFIG_ENV_IS_IN_SPI_FLASH=y
+CONFIG_BOOTP_SEND_HOSTNAME=y
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_V5L2_CACHE=y
+CONFIG_MMC=y
+CONFIG_FTSDC010=y
+CONFIG_FTSDC010_SDIO=y
+CONFIG_MTD_NOR_FLASH=y
+CONFIG_FLASH_CFI_DRIVER=y
+CONFIG_SYS_FLASH_USE_BUFFER_WRITE=y
+CONFIG_SYS_FLASH_CFI=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_FTMAC100=y
+CONFIG_BAUDRATE=38400
+CONFIG_SYS_NS16550=y
+CONFIG_SPI=y
+CONFIG_ATCSPI200_SPI=y
+# CONFIG_BINMAN_FDT is not set
diff --git a/configs/ae350_rv64_spl_xip_dynamic_defconfig b/configs/ae350_rv64_spl_xip_dynamic_defconfig
new file mode 100644
index 0000000000..67fcf880f4
--- /dev/null
+++ b/configs/ae350_rv64_spl_xip_dynamic_defconfig
@@ -0,0 +1,49 @@
+CONFIG_RISCV=y
+CONFIG_SYS_TEXT_BASE=0x01200000
+CONFIG_SYS_MALLOC_LEN=0x80000
+CONFIG_NR_DRAM_BANKS=2
+CONFIG_ENV_SECT_SIZE=0x1000
+CONFIG_DEFAULT_DEVICE_TREE="ae350_64"
+CONFIG_SPL_TEXT_BASE=0x80000000
+CONFIG_SPL_SYS_MALLOC_F_LEN=0x1800000
+CONFIG_SPL=y
+CONFIG_TARGET_ANDES45_AE350=y
+CONFIG_ARCH_RV64I=y
+CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_DYNAMIC=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman_d.dtsi"
+CONFIG_XIP=y
+CONFIG_DISTRO_DEFAULTS=y
+CONFIG_SYS_LOAD_ADDR=0x100000
+CONFIG_FIT=y
+CONFIG_SPL_LOAD_FIT_ADDRESS=0x88600000
+CONFIG_BOOTDELAY=3
+CONFIG_MISC_INIT_R=y
+CONFIG_SPL_BOARD_INIT=y
+CONFIG_SPL_CACHE=y
+CONFIG_SYS_PROMPT="RISC-V # "
+CONFIG_CMD_IMLS=y
+CONFIG_CMD_MMC=y
+CONFIG_CMD_SF_TEST=y
+# CONFIG_CMD_SETEXPR is not set
+CONFIG_BOOTP_PREFER_SERVERIP=y
+CONFIG_CMD_CACHE=y
+CONFIG_ENV_OVERWRITE=y
+CONFIG_ENV_IS_IN_SPI_FLASH=y
+CONFIG_BOOTP_SEND_HOSTNAME=y
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_V5L2_CACHE=y
+CONFIG_MMC=y
+CONFIG_FTSDC010=y
+CONFIG_FTSDC010_SDIO=y
+CONFIG_MTD_NOR_FLASH=y
+CONFIG_FLASH_CFI_DRIVER=y
+CONFIG_SYS_FLASH_USE_BUFFER_WRITE=y
+CONFIG_SYS_FLASH_CFI=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_FTMAC100=y
+CONFIG_BAUDRATE=38400
+CONFIG_SYS_NS16550=y
+CONFIG_SPI=y
+CONFIG_ATCSPI200_SPI=y
+# CONFIG_BINMAN_FDT is not set
diff --git a/configs/ae350_rv64_spl_xip_jump_defconfig b/configs/ae350_rv64_spl_xip_jump_defconfig
new file mode 100644
index 0000000000..22f4a13a8e
--- /dev/null
+++ b/configs/ae350_rv64_spl_xip_jump_defconfig
@@ -0,0 +1,49 @@
+CONFIG_RISCV=y
+CONFIG_SYS_TEXT_BASE=0x0200000
+CONFIG_SYS_MALLOC_LEN=0x80000
+CONFIG_NR_DRAM_BANKS=2
+CONFIG_ENV_SECT_SIZE=0x1000
+CONFIG_DEFAULT_DEVICE_TREE="ae350_64"
+CONFIG_SPL_TEXT_BASE=0x80000000
+CONFIG_SPL_SYS_MALLOC_F_LEN=0x1800000
+CONFIG_SPL=y
+CONFIG_TARGET_ANDES45_AE350=y
+CONFIG_ARCH_RV64I=y
+CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_JUMP=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman_j.dtsi"
+CONFIG_XIP=y
+CONFIG_DISTRO_DEFAULTS=y
+CONFIG_SYS_LOAD_ADDR=0x100000
+CONFIG_FIT=y
+CONFIG_SPL_LOAD_FIT_ADDRESS=0x88600000
+CONFIG_BOOTDELAY=3
+CONFIG_MISC_INIT_R=y
+CONFIG_SPL_BOARD_INIT=y
+CONFIG_SPL_CACHE=y
+CONFIG_SYS_PROMPT="RISC-V # "
+CONFIG_CMD_IMLS=y
+CONFIG_CMD_MMC=y
+CONFIG_CMD_SF_TEST=y
+# CONFIG_CMD_SETEXPR is not set
+CONFIG_BOOTP_PREFER_SERVERIP=y
+CONFIG_CMD_CACHE=y
+CONFIG_ENV_OVERWRITE=y
+CONFIG_ENV_IS_IN_SPI_FLASH=y
+CONFIG_BOOTP_SEND_HOSTNAME=y
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_V5L2_CACHE=y
+CONFIG_MMC=y
+CONFIG_FTSDC010=y
+CONFIG_FTSDC010_SDIO=y
+CONFIG_MTD_NOR_FLASH=y
+CONFIG_FLASH_CFI_DRIVER=y
+CONFIG_SYS_FLASH_USE_BUFFER_WRITE=y
+CONFIG_SYS_FLASH_CFI=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_FTMAC100=y
+CONFIG_BAUDRATE=38400
+CONFIG_SYS_NS16550=y
+CONFIG_SPI=y
+CONFIG_ATCSPI200_SPI=y
+# CONFIG_BINMAN_FDT is not set
diff --git a/configs/ae350_rv64_spl_xip_payload_defconfig b/configs/ae350_rv64_spl_xip_payload_defconfig
new file mode 100644
index 0000000000..5b6c731091
--- /dev/null
+++ b/configs/ae350_rv64_spl_xip_payload_defconfig
@@ -0,0 +1,49 @@
+CONFIG_RISCV=y
+CONFIG_SYS_TEXT_BASE=0x01200000
+CONFIG_SYS_MALLOC_LEN=0x80000
+CONFIG_NR_DRAM_BANKS=2
+CONFIG_ENV_SECT_SIZE=0x1000
+CONFIG_DEFAULT_DEVICE_TREE="ae350_64"
+CONFIG_SPL_TEXT_BASE=0x80000000
+CONFIG_SPL_SYS_MALLOC_F_LEN=0x1800000
+CONFIG_SPL=y
+CONFIG_TARGET_ANDES45_AE350=y
+CONFIG_ARCH_RV64I=y
+CONFIG_RISCV_SMODE=y
+CONFIG_SPL_LOAD_FIT_PAYLOAD=y
+CONFIG_SPL_LOAD_FIT_CONFIG="binman_p.dtsi"
+CONFIG_XIP=y
+CONFIG_DISTRO_DEFAULTS=y
+CONFIG_SYS_LOAD_ADDR=0x100000
+CONFIG_FIT=y
+CONFIG_SPL_LOAD_FIT_ADDRESS=0x88600000
+CONFIG_BOOTDELAY=3
+CONFIG_MISC_INIT_R=y
+CONFIG_SPL_BOARD_INIT=y
+CONFIG_SPL_CACHE=y
+CONFIG_SYS_PROMPT="RISC-V # "
+CONFIG_CMD_IMLS=y
+CONFIG_CMD_MMC=y
+CONFIG_CMD_SF_TEST=y
+# CONFIG_CMD_SETEXPR is not set
+CONFIG_BOOTP_PREFER_SERVERIP=y
+CONFIG_CMD_CACHE=y
+CONFIG_ENV_OVERWRITE=y
+CONFIG_ENV_IS_IN_SPI_FLASH=y
+CONFIG_BOOTP_SEND_HOSTNAME=y
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_V5L2_CACHE=y
+CONFIG_MMC=y
+CONFIG_FTSDC010=y
+CONFIG_FTSDC010_SDIO=y
+CONFIG_MTD_NOR_FLASH=y
+CONFIG_FLASH_CFI_DRIVER=y
+CONFIG_SYS_FLASH_USE_BUFFER_WRITE=y
+CONFIG_SYS_FLASH_CFI=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_FTMAC100=y
+CONFIG_BAUDRATE=38400
+CONFIG_SYS_NS16550=y
+CONFIG_SPI=y
+CONFIG_ATCSPI200_SPI=y
+# CONFIG_BINMAN_FDT is not set
diff --git a/configs/qilai_a45_spl_defconfig b/configs/qilai_a45_spl_defconfig
index de20d9d310..e54b58fb3e 100644
--- a/configs/qilai_a45_spl_defconfig
+++ b/configs/qilai_a45_spl_defconfig
@@ -1,20 +1,18 @@
 CONFIG_RISCV=y
-CONFIG_SYS_TEXT_BASE=0x01200000
-CONFIG_SYS_MALLOC_LEN=0x80000
+CONFIG_SYS_TEXT_BASE=0x401200000
+CONFIG_SYS_MALLOC_LEN=0x800000
 CONFIG_NR_DRAM_BANKS=2
 CONFIG_ENV_SECT_SIZE=0x1000
 CONFIG_DEFAULT_DEVICE_TREE="qilai_32"
-CONFIG_SPL_TEXT_BASE=0x800000
 CONFIG_SPL_SYS_MALLOC_F_LEN=0x100000
 CONFIG_SPL=y
-CONFIG_SYS_LOAD_ADDR=0x100000
 CONFIG_TARGET_ANDES45_QILAI=y
 CONFIG_RISCV_SMODE=y
 CONFIG_DISTRO_DEFAULTS=y
-CONFIG_PHYS_64BIT=y
+CONFIG_SYS_LOAD_ADDR=0x100000
 CONFIG_FIT=y
-CONFIG_SPL_LOAD_FIT_ADDRESS=0x00200000
-CONFIG_SYS_MONITOR_BASE=0x88000000
+CONFIG_SPL_LOAD_FIT_CONFIG="binman.dtsi"
+CONFIG_SPL_LOAD_FIT_ADDRESS=0x400200000
 CONFIG_BOOTDELAY=3
 CONFIG_MISC_INIT_R=y
 CONFIG_SPL_CACHE=y
@@ -26,7 +24,6 @@ CONFIG_CMD_SF_TEST=y
 CONFIG_BOOTP_PREFER_SERVERIP=y
 CONFIG_CMD_CACHE=y
 CONFIG_OF_BOARD=y
-CONFIG_OF_LIST="ae350_64"
 CONFIG_ENV_OVERWRITE=y
 CONFIG_ENV_IS_IN_SPI_FLASH=y
 CONFIG_BOOTP_SEND_HOSTNAME=y
@@ -35,7 +32,7 @@ CONFIG_V5L2_CACHE=y
 CONFIG_MMC=y
 CONFIG_FTSDC010=y
 CONFIG_FTSDC010_SDIO=y
-CONFIG_MTD_NOR_FLASH=y
+# CONFIG_MTD_NOR_FLASH is not set
 CONFIG_FLASH_CFI_DRIVER=y
 CONFIG_SYS_FLASH_USE_BUFFER_WRITE=y
 CONFIG_SYS_FLASH_CFI=y
diff --git a/configs/qilai_ax45_spl_defconfig b/configs/qilai_ax45_spl_defconfig
index 6b59708621..e3667943e3 100644
--- a/configs/qilai_ax45_spl_defconfig
+++ b/configs/qilai_ax45_spl_defconfig
@@ -1,19 +1,19 @@
 CONFIG_RISCV=y
-CONFIG_SYS_TEXT_BASE=0x01200000
-CONFIG_SYS_MALLOC_LEN=0x80000
+CONFIG_SYS_TEXT_BASE=0x401200000
+CONFIG_SYS_MALLOC_LEN=0x800000
 CONFIG_NR_DRAM_BANKS=2
 CONFIG_ENV_SECT_SIZE=0x1000
-CONFIG_DEFAULT_DEVICE_TREE="ae350_64"
+CONFIG_DEFAULT_DEVICE_TREE="qilai_64"
 CONFIG_SPL_SYS_MALLOC_F_LEN=0x100000
 CONFIG_SPL=y
-CONFIG_TARGET_ANDES45_AE350=y
+CONFIG_TARGET_ANDES45_QILAI=y
 CONFIG_ARCH_RV64I=y
 CONFIG_RISCV_SMODE=y
 CONFIG_DISTRO_DEFAULTS=y
 CONFIG_SYS_LOAD_ADDR=0x100000
 CONFIG_FIT=y
-CONFIG_SPL_LOAD_FIT_ADDRESS=0x00200000
-CONFIG_SYS_MONITOR_BASE=0x88000000
+CONFIG_SPL_LOAD_FIT_CONFIG="binman.dtsi"
+CONFIG_SPL_LOAD_FIT_ADDRESS=0x400200000
 CONFIG_BOOTDELAY=3
 CONFIG_MISC_INIT_R=y
 CONFIG_SPL_CACHE=y
@@ -33,7 +33,7 @@ CONFIG_V5L2_CACHE=y
 CONFIG_MMC=y
 CONFIG_FTSDC010=y
 CONFIG_FTSDC010_SDIO=y
-CONFIG_MTD_NOR_FLASH=y
+# CONFIG_MTD_NOR_FLASH is not set
 CONFIG_FLASH_CFI_DRIVER=y
 CONFIG_SYS_FLASH_USE_BUFFER_WRITE=y
 CONFIG_SYS_FLASH_CFI=y
diff --git a/include/configs/andes25-ae350.h b/include/configs/andes25-ae350.h
index 3903dcf3e9..9d5ae520a4 100644
--- a/include/configs/andes25-ae350.h
+++ b/include/configs/andes25-ae350.h
@@ -13,7 +13,11 @@
 #define CONFIG_SPL_BSS_MAX_SIZE		0x00100000
 
 #ifdef CONFIG_SPL_MMC
+#ifdef CONFIG_SPL_LOAD_FIT_NORMAL
 #define CONFIG_SPL_FS_LOAD_PAYLOAD_NAME		"u-boot.itb"
+#else
+#define CONFIG_SPL_FS_LOAD_PAYLOAD_NAME		"linux.itb"
+#endif
 #endif
 #endif
 
@@ -70,7 +74,7 @@
 #define CONFIG_SYS_NS16550_CLK		19660800
 
 /* Init Stack Pointer */
-#define CONFIG_SYS_INIT_SP_ADDR		(CONFIG_SYS_SDRAM_BASE + 0x1000000 - \
+#define CONFIG_SYS_INIT_SP_ADDR		(CONFIG_SYS_SDRAM_BASE + 0x4000000 - \
 					GENERATED_GBL_DATA_SIZE)
 
 /* use CFI framework */
@@ -133,11 +137,11 @@
 #include <config_distro_bootcmd.h>
 
 #define CONFIG_EXTRA_ENV_SETTINGS	\
-				"kernel_addr_r=0x00080000\0" \
+				"kernel_addr_r=0x00600000\0" \
 				"pxefile_addr_r=0x01f00000\0" \
 				"scriptaddr=0x01f00000\0" \
-				"fdt_addr_r=0x02000000\0" \
-				"ramdisk_addr_r=0x02800000\0" \
+				"fdt_addr_r=0x04600000\0" \
+				"ramdisk_addr_r=0x04800000\0" \
 				BOOTENV
 
 #endif /* __CONFIG_H */
diff --git a/include/configs/andes45-ae350.h b/include/configs/andes45-ae350.h
index 8b8b9ef066..7b23a9795e 100644
--- a/include/configs/andes45-ae350.h
+++ b/include/configs/andes45-ae350.h
@@ -15,7 +15,11 @@
 #define CONFIG_SPL_BSS_MAX_SIZE		0x00100000
 
 #ifdef CONFIG_SPL_MMC
+#ifdef CONFIG_SPL_LOAD_FIT_NORMAL
 #define CONFIG_SPL_FS_LOAD_PAYLOAD_NAME		"u-boot.itb"
+#else
+#define CONFIG_SPL_FS_LOAD_PAYLOAD_NAME		"linux.itb"
+#endif
 #endif
 #endif
 
@@ -74,7 +78,7 @@
 #define CONFIG_SYS_NS16550_CLK		19660800
 
 /* Init Stack Pointer */
-#define CONFIG_SYS_INIT_SP_ADDR		(CONFIG_SYS_SDRAM_BASE + 0x10000000 - \
+#define CONFIG_SYS_INIT_SP_ADDR		(CONFIG_SYS_SDRAM_BASE + 0x4000000 - \
 					GENERATED_GBL_DATA_SIZE)
 
 /* use CFI framework */
@@ -125,11 +129,11 @@
 #include <config_distro_bootcmd.h>
 
 #define CONFIG_EXTRA_ENV_SETTINGS	\
-				"kernel_addr_r=0x00080000\0" \
+				"kernel_addr_r=0x00600000\0" \
 				"pxefile_addr_r=0x01f00000\0" \
 				"scriptaddr=0x01f00000\0" \
-				"fdt_addr_r=0x02000000\0" \
-				"ramdisk_addr_r=0x02800000\0" \
+				"fdt_addr_r=0x04600000\0" \
+				"ramdisk_addr_r=0x04800000\0" \
 				BOOTENV
 
 #endif /* __CONFIG_H */
diff --git a/include/configs/andes45-qilai.h b/include/configs/andes45-qilai.h
index 8b8b9ef066..b94dc0c145 100644
--- a/include/configs/andes45-qilai.h
+++ b/include/configs/andes45-qilai.h
@@ -11,11 +11,18 @@
 
 #ifdef CONFIG_SPL
 #define CONFIG_SPL_MAX_SIZE		0x00100000
-#define CONFIG_SPL_BSS_START_ADDR	0x04000000
+#define CONFIG_SPL_BSS_START_ADDR	0x404000000
 #define CONFIG_SPL_BSS_MAX_SIZE		0x00100000
 
+#define CONFIG_SYS_SPL_MALLOC_START	0x8000000
+#define CONFIG_SYS_SPL_MALLOC_SIZE	0x2000000
+
 #ifdef CONFIG_SPL_MMC
+#ifdef CONFIG_SPL_LOAD_FIT_NORMAL
 #define CONFIG_SPL_FS_LOAD_PAYLOAD_NAME		"u-boot.itb"
+#else
+#define CONFIG_SPL_FS_LOAD_PAYLOAD_NAME		"linux.itb"
+#endif
 #endif
 #endif
 
@@ -52,12 +59,12 @@
 #define CONFIG_SYS_NONCACHED_MEMORY	SZ_1M
 
 /* DT blob (fdt) address */
-#define CONFIG_SYS_FDT_BASE		0x800f0000
+#define CONFIG_SYS_FDT_BASE		0x4800f0000
 
 /*
  * Physical Memory Map
  */
-#define PHYS_SDRAM_0	0x00000000		/* SDRAM Bank #1 */
+#define PHYS_SDRAM_0	0x400000000		/* SDRAM Bank #1 */
 #define PHYS_SDRAM_1	\
 	(PHYS_SDRAM_0 + PHYS_SDRAM_0_SIZE)	/* SDRAM Bank #2 */
 #define PHYS_SDRAM_0_SIZE	0x20000000	/* 512 MB */
@@ -74,7 +81,7 @@
 #define CONFIG_SYS_NS16550_CLK		19660800
 
 /* Init Stack Pointer */
-#define CONFIG_SYS_INIT_SP_ADDR		(CONFIG_SYS_SDRAM_BASE + 0x10000000 - \
+#define CONFIG_SYS_INIT_SP_ADDR		(CONFIG_SYS_SDRAM_BASE + 0x4000000 - \
 					GENERATED_GBL_DATA_SIZE)
 
 /* use CFI framework */
@@ -86,6 +93,7 @@
 #define PHYS_FLASH_1			0x88000000	/* BANK 0 */
 #define CONFIG_SYS_FLASH_BASE		PHYS_FLASH_1
 #define CONFIG_SYS_FLASH_BANKS_LIST	{ PHYS_FLASH_1, }
+#define CONFIG_SYS_MONITOR_BASE		PHYS_FLASH_1
 
 #define CONFIG_SYS_FLASH_ERASE_TOUT	120000	/* TO for Flash Erase (ms) */
 #define CONFIG_SYS_FLASH_WRITE_TOUT	500	/* TO for Flash Write (ms) */
@@ -125,11 +133,11 @@
 #include <config_distro_bootcmd.h>
 
 #define CONFIG_EXTRA_ENV_SETTINGS	\
-				"kernel_addr_r=0x00080000\0" \
-				"pxefile_addr_r=0x01f00000\0" \
-				"scriptaddr=0x01f00000\0" \
-				"fdt_addr_r=0x02000000\0" \
-				"ramdisk_addr_r=0x02800000\0" \
+				"kernel_addr_r=0x400600000\0" \
+				"pxefile_addr_r=0x401f00000\0" \
+				"scriptaddr=0x401f00000\0" \
+				"fdt_addr_r=0x404600000\0" \
+				"ramdisk_addr_r=0x404800000\0" \
 				BOOTENV
 
 #endif /* __CONFIG_H */
-- 
2.34.1

